### DISCUSSION TITLE
Monogame with Avalonia has "popping" on linux systems

### DISCUSSION THREAD

**Discussion by citruslee** [AvaloniaUI/Avalonia#14315] [General]
## Describe the bug
When I use Monogame in an Avalonia application on linux (latest Pop OS!), I am encountering some skipping in animation. I have circular motion set up inside the sample app

## To Reproduce

1. Download the attached [zip file of the engine](https://github.com/AvaloniaUI/Avalonia/files/14008021/HomunculusEngine.zip)
3. Open the project HomunculusEditor
4. Build and run
5. Observe the choppiness

## Expected behavior
A smooth circular motion

## Environment
Lenovo Legion 16IRX8
13th gen Core i7-13700HX
16gb RAM
RTX 4060 8gb VRAM mobile
Pop_OS! latest

- OS: Linux Pop_OS!
- Avalonia-Version: 11.0.2.0

## Additional context
Happens only on linux specifically. Monogame runs smooth without Avalonia, as you can see when running the engine itself.

**Comment by timunie** [comment]
please test nightly or at least 11.0.7. Also make sure only to measure performance in Release Mode.

**Comment by citruslee** [comment]
the thing is, it is bang on 60FPS. In both release and debug builds. But it is still choppy. I will try to do 11.0.7 avalonia then

**Comment by citruslee** [comment]
on 11.0.7 it is still choppy

**Comment by maxkatz6** [comment]
Significant problems with performance after a first glance:
1. Reflection is used on each frame to update key state
2. `Dispatcher.UIThread.Post(InvalidateVisual, DispatcherPriority.Normal)` instead of `TopLevel.GetTopLevel(this).RequestAnimationFrame(timespan => InvalidateVisual());` (timespan argument can also be used to simplify some Update logic)
3. Writeable bitmap is used for rendering (hopefully, Monogame allows rendering into a OpenGlControl or something).

In general, I am moving this to the discussion, as it doesn't look like an Avalonia bug.

**Reply by citruslee** [reply]
tried the mentioned things, removed completely the reflection, yet it still is choppy.

**Reply by thevortexcloud** [reply]
Have you taken a look at https://github.com/AvaloniaInside/AvaloniaInside.MonoGame? I am not seeing it being used in your code

**Reply by citruslee** [reply]
no I have not. Is it relevant to my issue?

**Reply by damian-666** [reply]
Its similar to yours but does not use directX shaders are transpiled from HLSL  (FX)  to OpenGL via  mojoshader, and its not good quality, old version of openGl, no multisampling.


The monogame inside has
 **performance issues  and its also copying to a writeable bitmap**.  the tabbed version is tricky.  the dockable branch with monogame is also not updated..  if you need both.
 
 . your code, seems simple and maybe you can get it working looking at the  wpf way.. you set up the device and all that..   **but you are not really sharing the surface and  drawing to it.**

**The monogame.inside it copies to a WriteableBitmap, which it seems yours does also but you should not have to i hope**


  **Also I see you create a new rendertarget every update.  you should not have to.. just on **resize.****

**alos you are using  timers and such and the resolution of locks and timers in Linux and windows is poor.. By default its +- 16ms.**

So its AvaloniaInside is  fine for me on windows but reported to  have performance issue really bad, on linux, and its also copying to a writeable bitmap.  the tabbed version is tricky if you get to that.    but the answers are in the fork ill link, i think , they might port over, the markstan one its really overkill. ,  you will see the 
you mgith just need a few changes.

see the render code in here.


but this what you started, with directX,  might be very close to the best way with monogame and windows.  

but you are copying to a writeable bitmap instead of sharing the buffer.  since its a directX render target i dont know if its going to work. 

So you can may draw directly to the rendertarget and the special device will put it on the screen.   I think you set it up right... but this is very difficult.

for wpf and monogaem there were two ways.   one used a bad pattern the drawable list..   the marc stan one, its is super complete version with the input and everything.   and its a 404 abandoned.. i have the only surviving fork adn it has a ton of advanced considerations, tabbed.. allocating, reusing the backing store ( allocate the windows for tabs , allocate the desktop if they can float. and tear off..   that is the least memory you need.      

Locks and timers are slow and the producer consumer pattern is the best.. 


    MonoGame.WpfCore

this can take foreover,i just going to give what i had working. in wpf.
https://github.com/damian-666/monogame-framework-wpfinterop    the most complete but abandoned and leaks.  it might show yhou how to do certain things in dx.  I updated it to netcore 6 at least.. it builds and runs.  


https://github.com/damian-666/MonoGame.WpfCore   this is the abandoned craftwork games one, that was broken with two unmerged PRs.. those are in my fork and now it works..  it shows the part that  has one rendertarget adn creates it on every dray and doesnt cpy to writeeable bitmap.

And it doesnt use the Game class.


 if (!_isFirstLoad && CanBeginDraw())
 {
     try
     {
         _direct3DImage.Lock();

         if (_renderTarget == null)
             _renderTarget = CreateRenderTarget();

         if (_renderTarget != null)
         {
             GraphicsDevice.SetRenderTarget(_renderTarget);
             SetViewport();

             _viewModel?.Update(_gameTime);
             _viewModel?.Draw(_gameTime);

             GraphicsDevice.Flush();
             _direct3DImage.AddDirtyRect(new Int32Rect(0, 0, (int)ActualWidth, (int)ActualHeight));
         }
     }
     finally
     {
         _direct3DImage.Unlock();
         GraphicsDevice.SetRenderTarget(null);
         _viewModel?.AfterRender();
     }
 }


in general tabbed dialogs 
So you can look at BepuPhysics2 and see his demo. .that has a transpiler to openGL with shaderconductor.  it uses openTK.. openTK to avalonia is abandoned, last i checked.. it was not updtaed to the latest.. .     But you have figured out how to hook into DirectX, so that is ideal. 


I use the producer consumer pattern and a lockless approach.  so does bepu . i use my own timers.   drawing 3d monogame in adroid with xamarin , also i could nto use locks.


again stride is working on this..   You may have the option of looking at Bepuphysis and opentk , and not using monogame ..  and his special renderer...    or using monogame with opgl and the KNI branch.. but noone is testing that branch.

sorry about the typso... thats all i have . i never  decided to put mie on hold, stay withh wpf  and im on another project.  

i thin you should tryu to updaet teh render target and not ussg a wrteable bitmap... or transpile to opengl... 

there are lot of branches of somee of thiese but people are not sharing annd merging.. 

**hope it helps.. i spend a long time just trying to share beause i spend montsh on this.,..i konw Avalonia + Monoame is probably the idea  3d IDE mvoign forward because Stride has an ECS and is bloated**..
